From: Simon McVittie <smcv@debian.org>
Date: Wed, 25 Mar 2020 14:02:29 +0000
Subject: tests: Skip more tests if FUSE doesn't work

As with test-doc-portal.c, some CI and automated build environments
are locked-down to forbid FUSE for security or robustness reasons.
We can't run this test there either.

Capability detection recycled from flatpak.

Signed-off-by: Simon McVittie <smcv@debian.org>
---
 tests/Makefile.am.inc       |  1 +
 tests/libtest.sh            | 47 +++++++++++++++++++++++++++++++++++++++++++++
 tests/test-document-fuse.sh | 16 ++++-----------
 3 files changed, 52 insertions(+), 12 deletions(-)
 create mode 100644 tests/libtest.sh

diff --git a/tests/Makefile.am.inc b/tests/Makefile.am.inc
index 981c8d5..9d4f31b 100644
--- a/tests/Makefile.am.inc
+++ b/tests/Makefile.am.inc
@@ -109,6 +109,7 @@ dist_installed_test_portals_DATA = tests/portals/test.portal
 endif
 
 dist_installed_test_data = \
+	tests/libtest.sh \
 	tests/session.conf.in \
 	$(NULL)
 
diff --git a/tests/libtest.sh b/tests/libtest.sh
new file mode 100644
index 0000000..a3245e9
--- /dev/null
+++ b/tests/libtest.sh
@@ -0,0 +1,47 @@
+# Source library for shell script tests
+#
+# Copyright (C) 2016 Alexander Larsson <alexl@redhat.com>
+# Copyright (C) 2011 Colin Walters <walters@verbum.org>
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+# Boston, MA 02111-1307, USA.
+
+if [ -n "${G_TEST_SRCDIR:-}" ]; then
+    test_srcdir="${G_TEST_SRCDIR}"
+else
+    test_srcdir=$(realpath $(dirname $0))
+fi
+
+if [ -n "${G_TEST_BUILDDIR:-}" ]; then
+    test_builddir="${G_TEST_BUILDDIR}"
+else
+    test_builddir=$(realpath $(dirname $0))
+fi
+
+# Use to skip all of these tests
+skip() {
+    echo "1..0 # SKIP" "$@"
+    exit 0
+}
+
+skip_without_fuse () {
+    fusermount --version >/dev/null 2>&1 || skip "no fusermount"
+
+    capsh --print | grep -q 'Bounding set.*[^a-z]cap_sys_admin' || \
+        skip "No cap_sys_admin in bounding set, can't use FUSE"
+
+    [ -w /dev/fuse ] || skip "no write access to /dev/fuse"
+    [ -e /etc/mtab ] || skip "no /etc/mtab"
+}
diff --git a/tests/test-document-fuse.sh b/tests/test-document-fuse.sh
index fea0245..bea845e 100755
--- a/tests/test-document-fuse.sh
+++ b/tests/test-document-fuse.sh
@@ -1,20 +1,12 @@
 #!/bin/bash
 
-echo "1..2"
-
 set -e
 
-if [ -n "${G_TEST_SRCDIR:-}" ]; then
-    test_srcdir="${G_TEST_SRCDIR}"
-else
-    test_srcdir=$(realpath $(dirname $0))
-fi
+. "$(dirname "$0")/libtest.sh"
 
-if [ -n "${G_TEST_BUILDDIR:-}" ]; then
-    test_builddir="${G_TEST_BUILDDIR}"
-else
-    test_builddir=$(realpath $(dirname $0))
-fi
+skip_without_fuse
+
+echo "1..2"
 
 export TEST_DATA_DIR=`mktemp -d /tmp/xdp-XXXXXX`
 mkdir -p ${TEST_DATA_DIR}/home
